plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -443,6 +473,58 @@ private static <K, V> Map<K, V> getAll(Cache<K, AtomicReference<V>> cache, Itera
         return result.buildOrThrow();
     }
 
+    private static <K, V> Map<K, V> getAll(
+            Cache<K, AtomicReference<V>> cache,
+            Iterable<K> keys,
+            Function<Collection<K>, Map<K, V>> bulkLoader,
+            Predicate<V> shouldReload,
+            BiFunction<V, V, V> merger)
+    {
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        Map<K, AtomicReference<V>> valueHolders = new HashMap<>();
+
+        ImmutableList.Builder<K> missingKeysBuilder = ImmutableList.builder();
+        keys.forEach(key -> {
+            // make sure the value holder is retrieved before the new values are loaded
+            // so that in case of invalidation we will not set the stale value
+            AtomicReference<V> currentValueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+            valueHolders.put(key, currentValueHolder);
```

Comment:
value holders are used for missing keys only, so you can move  this line into `else`, next to `missingKeysBuilder.add(key);` 
(rename `valueHolders` when doing so)

Reply:
also, you could have just one collection, instead of two

```
// Key to value holder (either empty, or with insufficient information)
Map<K, AtomicReference<V>> toLoad = new HashMap<>();
```